File: 0934-shortest-bridge/0934-shortest-bridge.py
class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        n = len(grid)
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        q = deque()
        firstX = -1
        firstY = -1
        for i in range(n):
            for j in range(n):
                if grid[i][j] == 1:
                    firstX = i
                    firstY = j
                    break
            if firstX != -1:
                break
        self.dfs(grid, firstX, firstY, n, q)
        distance = 0
        while q:
            size = len(q)
            for _ in range(size):
                x, y = q.popleft()
                for k in range(4):
                    curX = x + dx[k]
                    curY = y + dy[k]
                    if 0 <= curX < n and 0 <= curY < n:
                        if grid[curX][curY] == 1:
                            return distance
                        elif grid[curX][curY] == 0:
                            q.append((curX, curY))
                            grid[curX][curY] = -1
            distance += 1
        return distance
    def dfs(self, grid, x, y, n, q):
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        grid[x][y] = 2
        q.append((x, y))
        for k in range(4):
            curX = x + dx[k]
            curY = y + dy[k]
            if 0 <= curX < n and 0 <= curY < n and grid[curX][curY] == 1:
                self.dfs(grid, curX, curY, n, q)

File: 0934-shortest-bridge/0934-shortest-bridge.java
class Solution {
    int[] dx = {1, -1, 0, 0};
    int[] dy = {0, 0, 1, -1};
    Queue<int[]> q = new LinkedList<>();
    public int shortestBridge(int[][] grid) {
        int n = grid.length;
        int firstX = -1, firstY = -1;
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == 1){
                    firstX = i;
                    firstY = j;
                    break;
                }
            }
        }
        dfs(grid, firstX, firstY, n);
        int distance = 0;
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++) {
                int[] p = q.poll();
                int x = p[0];
                int y = p[1];
                for(int k = 0; k < 4; k++) {
                    int curX = x + dx[k];
                    int curY = y + dy[k];
                    if(0 <= curX && curX < n && 0 <= curY && curY < n) {
                        if(grid[curX][curY] == 1) {
                            return distance;
                        } else if(grid[curX][curY] == 0) {
                            q.offer(new int[]{curX, curY});
                            grid[curX][curY] = -1;
                        }
                    }
                }
            }
            distance++;
        }
        return distance;
    }
    void dfs(int[][] grid, int x, int y, int n){
        grid[x][y] = 2;
        q.offer(new int[]{x, y});
        for(int k=0; k<4; k++){
            int curX = x+dx[k];
            int curY = y+dy[k];
            if(0<=curX && curX<n && 0<=curY && curY<n && grid[curX][curY]==1){
                dfs(grid, curX, curY, n);
            }
        }
    }
    
}